Add ViViT

jegork opened the pull request:
# What does this PR do?

Fixes #15666. Reopening #20441, as I have missed the comments provided by @amyeroberts so the issue was closed by the bot. 

Add Video Vision Transformer to transformers. This PR implements a spacetime version of the Video Vision Transformer from the original paper. 

I have provided the model weights here https://huggingface.co/jegormeister/vivit-b-16x2-kinetics400 
I will try to add Factorised Encoder version later on (these are the two versions that authors provide weight for).


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? https://github.com/huggingface/transformers/issues/15666
- [x] Did you make sure to update the documentation with your changes? I have added the documentation, but I have troubles testing it as I couldn't run the preview command of the doc-builder, so if someone has the possibility to run and check it, I will be really grateful!
- [x] Did you write any new necessary tests? WIP


## Who can review?

@amyeroberts provided the last suggestions to the closed PR, so I hope you can review this one. Thanks!

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- albert, bert, xlm: @LysandreJik
- blenderbot, bart, marian, pegasus, encoderdecoder,  t5: @patrickvonplaten, @patil-suraj
- longformer, reformer, transfoxl, xlnet: @patrickvonplaten
- fsmt: @stas00
- funnel: @sgugger
- gpt2: @patrickvonplaten, @LysandreJik
- rag: @patrickvonplaten, @lhoestq
- tensorflow: @LysandreJik

Library:

- benchmarks: @patrickvonplaten
- deepspeed: @stas00
- ray/raytune: @richardliaw, @amogkam
- text generation: @patrickvonplaten
- tokenizers: @n1t0, @LysandreJik
- trainer: @sgugger
- pipelines: @LysandreJik

Documentation: @sgugger

HF projects:

- datasets: [different repo](https://github.com/huggingface/datasets)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Examples:

- maintained examples (not research project or legacy): @sgugger, @patil-suraj
- research_projects/bert-loses-patience: @JetRunner
- research_projects/distillation: @VictorSanh

 -->

jegork committed:
Add model

jegork committed:
Add model

jegork committed:
Add ability to get classification head weights

jegork committed:
Add docs

jegork committed:
Add docs

jegork committed:
Add imports to __init__.py

jegork committed:
Run style

jegork committed:
Run style

jegork committed:
Fix imports and add mdx doc

jegork committed:
Fix imports and add mdx doc

jegork committed:
Run style

jegork committed:
Fix copyright

jegork committed:
Fix config docstring

jegork committed:
Remove imports of ViViTLayer and load_tf_weights_in_vivit

jegork committed:
Remove FeatureExtractor and replace with ImageProcessor everywhere

jegork committed:
Remove FeatureExtractor and replace with ImageProcessor everywhere

jegork committed:
Remove ViViTForPreTraining from vivit.mdx

jegork committed:
Change ViViT -> Vivit everywhere

jegork committed:
Change ViViT -> Vivit everywhere

jegork committed:
Add model_doc to _toctree.yml

jegork committed:
Replace tuples with lists in arguments of VivitConfig

jegork committed:
Rename patch_size to tubelet_size in TubeletEmbeddings

jegork committed:
Fix checkpoint names

jegork committed:
Add tests

jegork committed:
Remove unused num_frames

jegork committed:
Fix imports for VivitImageProcessor

jegork committed:
Minor fixes

jegork committed:
Decrease number of frames in VivitModelTester from 32 to 16

jegork committed:
Decrease number of frames in VivitModelTester from 16 to 8

jegork committed:
Add initialization for pos embeddings

jegork committed:
Rename Vivit -> ViViT in some places

jegork committed:
Rename Vivit -> ViViT in some places

jegork committed:
Fix docstring and formatting

jegork committed:
Rename TubeletEmbeddings -> VivitTubeletEmbeddings

jegork committed:
Remove load_tf_weights_in_vivit

jegork committed:
Change checkpoint name

jegork committed:
Remove Vivit _TOKENIZER_FOR_DOC

jegork committed:
Fix

jegork committed:
Fix VivitTubeletEmbeddings and pass config object as parameter

jegork committed:
Use image_size and num_frames instead of video_size

jegork committed:
Change conversion script and fix differences with the orig implementation

jegork committed:
Fix docstrings

jegork committed:
Add attention head pruning

jegork committed:
Run style and fixup

jegork committed:
Fix tests

jegork committed:
Add ViViT to video_classification.mdx

jegork committed:
Save processor in conversion script

jegork committed:
Fix

jegork committed:
Add image processor test

jegork committed:
Run fixup and style

jegork committed:
Run fix-copies

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_22518). All of your documentation changes will be reflected on that endpoint.

jegork commented:
@amyeroberts thank you for your comments in the previous PR! I have addressed your suggestions and also added the image processor test. 

However, I ran `make style`, however the pipeline was failing at check_code_quality. Therefore I've updated the testing dependencies which lead to black updating. But! When I run `make style` again, it gives the following output:

```sh
reformatted transformers/examples/research_projects/deebert/src/modeling_highway_bert.py
reformatted transformers/examples/research_projects/movement-pruning/emmental/modeling_bert_masked.py
reformatted transformers/src/transformers/models/reformer/modeling_reformer.py
reformatted transformers/src/transformers/models/vivit/modeling_vivit.py
reformatted transformers/tests/models/vivit/test_image_processing_vivit.py
```

So it fixes not only the files from this PR, but also the already existing ones. Therefore I have a question: should I only pus reformatted files from this PR or all?

amyeroberts commented:
@jegork The files listed e.g. `transformers/examples/research_projects/deebert/src/modeling_highway_bert.py` should have the most recent formatting applied and shouldn't need to be updated with this PR. 

Could you rebase on main, make sure the most recent formatting packages are installed using `pip install -e .[quality]` and try `make style` again?

amyeroberts commented on tests/models/vivit/test_modeling_vivit.py:
```diff
@@ -0,0 +1,350 @@
+# coding=utf-8
+# Copyright 2022 The HuggingFace Inc. team. All rights reserved.
```
```suggestion
# Copyright 2023 The HuggingFace Inc. team. All rights reserved.
```

amyeroberts reviewed (commented):
Very nice! ❤️  Mostly small nit comments. Only necessary update before merging is removing the dependence on `decord` in the doc examples. 

Let me know if the additional diffs from `make style` are still appearing after updating styling packages and rebasing.

jegork committed:
Update tests/models/vivit/test_modeling_vivit.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

jegork committed:
Update tests/models/vivit/test_modeling_vivit.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

jegork committed:
Update src/transformers/models/vivit/modeling_vivit.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

jegork committed:
Use PyAV instead of decord

jegork committed:
Add unittest.skip

jegork committed:
Run style

jegork commented on tests/models/vivit/test_image_processing_vivit.py:
```diff
@@ -0,0 +1,218 @@
+        self.assertEqual(image_processor.size, {"shortest_edge": 42})
+        self.assertEqual(image_processor.crop_size, {"height": 84, "width": 84})
+
+    def test_batch_feature(self):
```
This part was copied from the VideoMAE tests as you suggested, but I have added the decorator

jegork commented:
@amyeroberts Thanks for your comments. I have addressed everything. However, I still get the same problems with `make style` 
I did `git fetch upstream`, then `git rebase upstream/main`, `pip install -e ".[quality]"` after which I ran `make style`

Which resulted in the following output at the end:
```sh
black examples tests src utils setup.py
Skipping .ipynb files as Jupyter dependencies are not installed.
You can fix this by running ``pip install "black[jupyter]"``
reformatted /Users/jegorkitskerkin/Documents/projects/transformers/examples/research_projects/deebert/src/modeling_highway_bert.py
reformatted /Users/jegorkitskerkin/Documents/projects/transformers/examples/research_projects/movement-pruning/emmental/modeling_bert_masked.py
reformatted /Users/jegorkitskerkin/Documents/projects/transformers/src/transformers/models/vivit/modeling_vivit.py
reformatted /Users/jegorkitskerkin/Documents/projects/transformers/src/transformers/models/reformer/modeling_reformer.py
reformatted /Users/jegorkitskerkin/Documents/projects/transformers/tests/models/vivit/test_image_processing_vivit.py

All done! ✨ 🍰 ✨
5 files reformatted, 2380 files left unchanged.
ruff examples tests src utils setup.py --fix
/Library/Developer/CommandLineTools/usr/bin/make autogenerate_code
running deps_table_update
updating src/transformers/dependency_versions_table.py
/Library/Developer/CommandLineTools/usr/bin/make extra_style_checks
python utils/custom_init_isort.py
python utils/sort_auto_mappings.py
doc-builder style src/transformers docs/source --max_len 119 --path_to_docs docs/source
Overwriting content of src/transformers/models/vivit/modeling_vivit.py.
Cleaned 1 files!
python utils/check_doc_toc.py --fix_and_overwrite
```

As you can see, the same unrelated-to-this-PR-files are getting formatted

jegork committed:
Remove unneeded test

amyeroberts commented on src/transformers/models/vivit/configuration_vivit.py:
```diff
@@ -0,0 +1,122 @@
+            The non-linear activation function (function or string) in the encoder and pooler. If string, `"gelu"`,
+            `"relu"`, `"selu"`, `"gelu_fast"` and `"gelu_new"` are supported.
+        hidden_dropout_prob (`float`, *optional*, defaults to 0.0):
+            The dropout probabilitiy for all fully connected layers in the embeddings, encoder, and pooler.
```
```suggestion
            The dropout probability for all fully connected layers in the embeddings, encoder, and pooler.
```

amyeroberts commented on src/transformers/models/vivit/modeling_vivit.py:
```diff
@@ -0,0 +1,734 @@
+        self.layernorm_after = nn.LayerNorm(config.hidden_size, eps=config.layer_norm_eps)
+
+    def forward(self, hidden_states, head_mask=None, output_attentions=False):
+        self_attention_outputs = self.attention(
+            self.layernorm_before(hidden_states),  # in Vivit, layernorm is applied before self-attention
+            head_mask,
+            output_attentions=output_attentions,
+        )
+        attention_output = self_attention_outputs[0]
+        outputs = self_attention_outputs[1:]  # add self attentions if we output attention weights
```
small nit: putting the comments on the line above the relevant piece of code

```suggestion
        self_attention_outputs = self.attention(
            # in Vivit, layernorm is applied before self-attention
            self.layernorm_before(hidden_states),
            head_mask,
            output_attentions=output_attentions,
        )
        attention_output = self_attention_outputs[0]
        # add self attentions if we output attention weights
        outputs = self_attention_outputs[1:]
```

amyeroberts commented on src/transformers/models/vivit/modeling_vivit.py:
```diff
@@ -0,0 +1,734 @@
+        ... )
+        >>> container = av.open(file_path)
+
+        >>> # sample 16 frames
+        >>> indices = sample_frame_indices(clip_len=32, frame_sample_rate=1, seg_len=len(videoreader))
+        >>> video = videoreader.get_batch(indices).asnumpy()
```
```suggestion
        >>> # sample 32 frames
        >>> indices = sample_frame_indices(clip_len=32, frame_sample_rate=1, seg_len=len(videoreader))
        >>> video = videoreader.get_batch(indices).asnumpy()
```

amyeroberts commented on src/transformers/models/vivit/modeling_vivit.py:
```diff
@@ -0,0 +1,734 @@
+        ... )
+        >>> container = av.open(file_path)
+
+        >>> # sample 16 frames
+        >>> indices = sample_frame_indices(clip_len=32, frame_sample_rate=1, seg_len=len(videoreader))
+        >>> video = videoreader.get_batch(indices).asnumpy()
```
```suggestion
        >>> # sample 32 frames
        >>> indices = sample_frame_indices(clip_len=32, frame_sample_rate=1, seg_len=len(videoreader))
        >>> video = videoreader.get_batch(indices).asnumpy()
```

amyeroberts commented on src/transformers/models/vivit/modeling_vivit.py:
```diff
@@ -0,0 +1,734 @@
+        >>> indices = sample_frame_indices(clip_len=32, frame_sample_rate=1, seg_len=len(videoreader))
+        >>> video = videoreader.get_batch(indices).asnumpy()
+
+        >>> feature_extractor = VivitImageProcessor.from_pretrained("google/vivit-b-16x2-kinetics400")
+        >>> model = VivitForVideoClassification.from_pretrained("google/vivit-b-16x2-kinetics400")
+
+        >>> inputs = feature_extractor(list(video), return_tensors="pt")
```
```suggestion
        >>> image_processor = VivitImageProcessor.from_pretrained("google/vivit-b-16x2-kinetics400")
        >>> model = VivitForVideoClassification.from_pretrained("google/vivit-b-16x2-kinetics400")

        >>> inputs = image_processor(list(video), return_tensors="pt")
```

amyeroberts commented on src/transformers/models/vivit/image_processing_vivit.py:
```diff
@@ -0,0 +1,407 @@
+            in the `preprocess` method.
+        do_zero_centering (`bool`, *optional*, defaults to `True`):
+            Whether to rescale the image to values between [-1 - 1] instead of [0 - 1]. Can be overriden by the
+            `do_zero_centering` in the `preprocess` method..
```
```suggestion
            `do_zero_centering` in the `preprocess` method.
```

amyeroberts commented on src/transformers/models/vivit/image_processing_vivit.py:
```diff
@@ -0,0 +1,407 @@
+            data_format (`str` or `ChannelDimension`, *optional*):
+                The channel dimension format of the image. If not provided, it will be the same as the input image.
+        """
+        if offset:
+            rescaled_image = (image - 127.5) * scale
+            if data_format is not None:
+                rescaled_image = to_channel_dimension_format(rescaled_image, data_format)
+            rescaled_image = rescaled_image.astype(np.float32)
+        else:
+            rescaled_image = rescale(image, scale=scale, data_format=data_format, **kwargs)
+
+        return rescale(image, scale=scale, data_format=data_format, **kwargs)
```
This function doesn't shift the image value between [-1, 1] if offset is `True` it creates `rescaled_image` and the doesn't use it. 

```suggestion
        image = image.astype(np.float32)        
        if offset:
        	image = image - 127.5
        return rescale(image, scale=scale, data_format=data_format, **kwargs)
```

amyeroberts commented on src/transformers/models/vivit/image_processing_vivit.py:
```diff
@@ -0,0 +1,407 @@
+        **kwargs,
+    ):
+        """
+        Rescale an image by a scale factor. image = image * scale.
```
```suggestion
                Rescale an image by a scale factor. 
        
        If offset is `True`, image scaled between [-1, 1]: image = (image - 127.5) * scale. 
        If offset is `False`, image scaled between [0, 1]: image = image * scale
```

amyeroberts commented on src/transformers/models/vivit/convert_vivit_flax_to_pytorch.py:
```diff
@@ -0,0 +1,236 @@
+
+    video = prepare_video()
+    inputs = extractor(video, return_tensors="pt")
+    torch.save(inputs, "video_processed.pt")
```
Why save the inputs out?

jegork replied:
@amyeroberts my bad, forgot to remove it after debugging. Fixed

amyeroberts reviewed (commented):
Thanks for adding this model! 

Overall very good PR - just a few small nits here and there. Only thing to be addressed is the `rescale` logic in the image processor. 

I'm not sure why some other files have changed if you've rebased and have the most recent formatting settings. There's only a small number of differences and they actually improve the formatting, so let's leave them be.

jegork committed:
Update docs/source/en/model_doc/vivit.mdx

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

jegork committed:
Update src/transformers/models/vivit/configuration_vivit.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

jegork committed:
Update src/transformers/models/vivit/modeling_vivit.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

jegork committed:
Update src/transformers/models/vivit/image_processing_vivit.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

jegork committed:
Update src/transformers/models/vivit/modeling_vivit.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

jegork committed:
Update src/transformers/models/vivit/modeling_vivit.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

jegork committed:
Update src/transformers/models/vivit/image_processing_vivit.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

jegork committed:
Update src/transformers/models/vivit/modeling_vivit.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

jegork committed:
Update src/transformers/models/vivit/image_processing_vivit.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

jegork committed:
Run make style

jegork committed:
Remove inputs save

jegork committed:
Fix image processor

jegork committed:
Fix

jegork committed:
Run `make style`

jegork commented:
@amyeroberts yep, everything works and passes locally

amyeroberts commented:
@jegork Thanks for confirming. I'm going to rerun CircleCI in case there was just some transient issue with the run. If it persists we can dig a bit more into it.

amyeroberts commented on tests/models/vivit/test_modeling_vivit.py:
```diff
@@ -0,0 +1,350 @@
+        is_training=True,
+        use_labels=True,
+        num_labels=10,
+        image_size=224,
+        num_frames=8,  # decreased, because default 32 takes too much RAM at inference
+        tubelet_size=[2, 16, 16],
+        num_channels=3,
+        hidden_size=768,
+        num_hidden_layers=12,
+        num_attention_heads=12,
+        intermediate_size=3072,
```
One thing to note here is that many of these values are very large and will cause the tests to run slowly. Could you update `image_size`, `num_frames`, `tubelete_size`, `hidden_size`, `num_attention_heads`, `num_hidden_layers` and `intermediate_size` to all be as small as possible? You can look at other model's test files to find typical values to use.

jegork replied:
thanks! i've decreased those values and the tests pass now!

adildahlan commented:
Thanks @amyeroberts and @jegork for working on this, we look forward to using the ViVit model!

jegork committed:
Decrease parameters of VivitModelTester

jegork committed:
Decrease tubelet size

jegork committed:
Rename vivit.mdx

amyeroberts commented on src/transformers/models/vivit/image_processing_vivit.py:
```diff
@@ -0,0 +1,404 @@
+        do_rescale: bool = True,
+        rescale_factor: Union[int, float] = 1 / 255,
+        offset: bool = True,
+        do_normalize: bool = False,
```
Just double checking this should be `False`. I'm sure it's OK, but unusual

jegork replied:
thanks for pointing it out! i've double checked the original code and it should've been `True`

amyeroberts commented on src/transformers/models/vivit/image_processing_vivit.py:
```diff
@@ -0,0 +1,404 @@
+        **kwargs,
+    ):
+        """
+                Rescale an image by a scale factor.
```
```suggestion
        Rescale an image by a scale factor.
```

amyeroberts commented on src/transformers/models/vivit/image_processing_vivit.py:
```diff
@@ -0,0 +1,404 @@
+        """
+        image = image.astype(np.float32)
+        if offset:
+            image = image - 127.5
```
```suggestion
            image = image - (scale / 2)
```

amyeroberts commented on src/transformers/models/vivit/image_processing_vivit.py:
```diff
@@ -0,0 +1,404 @@
+
+        Args:
+            videos (`ImageInput`):
+                Image to preprocess.
```
```suggestion
                Video frames to preprocess.
```

amyeroberts reviewed (approved):
Thanks for all the work adding this model! 

Just some nits to address and then we're good to merge!

jegork committed:
Update src/transformers/models/vivit/image_processing_vivit.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

jegork committed:
Update src/transformers/models/vivit/image_processing_vivit.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

jegork committed:
Update src/transformers/models/vivit/image_processing_vivit.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

jegork committed:
Fix default values in image_processing_vivit.py

amyeroberts commented:
@jegork Is the PR OK to merge? Or are there any other commits you'd like to push before I press the big green button? 🟢

jegork commented:
@amyeroberts I think it's ready to be merged. Thanks for your help!

NielsRogge commented:
Hi @jegork congrats on your amazing contribution!

is it ok if we transfer the ViViT checkpoints to the `google` organization on the hub? (assuming they are officially released checkpoints by Google)

jegork commented:
Hey @NielsRogge, thanks!

Sure
